File: C#/base/collections/CollectionInterfaces.cs
/**
 * Investigate the interfaces that make up collections in the .NET Framework
 * Sources: [C# 7.0 In a Nutshell: Page 301-313]
 * Author: Andrew Jarombek
 * Date: 7/29/2019
 */

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using static System.Diagnostics.Debug;

namespace collections
{
    /// <summary>
    /// My custom implementation of a List that is enumerable.  It has an inner Enumerator class.
    /// </summary>
    /// <typeparam name="T">Generic type for the internal list.</typeparam>
    class AndyList<T> : IEnumerable<T>, IEnumerable
    {
        private readonly List<T> internalList;

        /// <summary>
        /// Constructor for a list which takes any generic enumerable object as an argument.  This object is converted
        /// to a list structure.
        /// </summary>
        /// <param name="enumerable">An object which implements IEnumerable&lt;T&gt;</param>
        public AndyList(IEnumerable<T> enumerable)
        {
            internalList = enumerable.ToList();
        }

        /// <summary>
        /// Constructor for a list which takes any non-generic enumerable object as an argument.  This object is
        /// converted to a list structure.
        /// </summary>
        /// <param name="enumerable">An object which implements IEnumerable</param>
        public AndyList(IEnumerable enumerable)
        {
            internalList = enumerable.Cast<T>().ToList();
        }
        
        /// <summary>
        /// Get the Enumerator inner class which implements IEnumerator&lt;T&gt;.  The generic version of
        /// GetEnumerator() is implemented explicitly.
        /// </summary>
        /// <returns>An enumerator that implements the IEnumerator&lt;T&gt; interface methods.</returns>
        public IEnumerator<T> GetEnumerator()
        {
            return new Enumerator(internalList);
        }

        /// <summary>
        /// Get the Enumerator inner class which implements IEnumerator.  This is the non-generic version of
        /// GetEnumerator().  Per convention [pg. 307], the non-generic version is implemented implicitly. 
        /// </summary>
        /// <returns>An enumerator that implements the IEnumerator interface methods.</returns>
        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
        
        /// <summary>
        /// An enumerator which is used to iterate over the AndyList class.
        /// </summary>
        public struct Enumerator : IEnumerator<T>, IEnumerator, IDisposable
        {
            private readonly List<T> _list;
            private int _index;
            public T Current { get; private set; }

            /// <summary>
            /// Construct an enumerator for iterating over a list.  The list should be the one held internally
            /// inside the AndyList object instance.
            /// </summary>
            /// <param name="list">The list to iterate over.</param>
            internal Enumerator(List<T> list)
            {
                _list = list;
                _index = 0;
                Current = default(T);
            }
            
            /// <summary>
            /// Move to the next index in the list (if there is another index to navigate to).
            /// </summary>
            /// <returns>
            /// <code>true</code> if the next index is within the bounds of the list, <code>false</code> otherwise.
            /// </returns>
            public bool MoveNext()
            {
                if (_index >= _list.Count)
                {
                    _index = _list.Count + 1;
                    Current = default(T);
                    return false;
                }
                else
                {
                    Current = _list[_index];
                    _index++;
                    return true;
                }
            }

            /// <summary>
            /// Reset the current index in the list to the start of the list and the current value to the generic
            /// type's default value.
            /// </summary>
            public void Reset()
            {
                _index = 0;
                Current = default(T);
            }

            /// <summary>
            /// Retrieve the value at the index the enumerator is currently pointing to.
            /// </summary>
            object IEnumerator.Current => Current;

            /// <summary>
            /// There are no resources to dispose when done iterating over a list, so this method is empty.
            /// </summary>
            public void Dispose() {}
        }
    }
    
    public static class CollectionInterfaces
    {
        public static void Execute()
        {
            // Collections implement the IEnumerable and IEnumerable<T> interfaces.  They specify a single method
            // GetEnumerator(), which returns a class that implements IEnumerable and IEnumerable<T>.  IEnumerable
            // specifies three methods - MoveNext(), Current(), and Reset().
            // Java = [https://bit.ly/2YnRP7Q]
            var list = new List<int> {2, 4, 6};
            IEnumerator<int> enumerator = list.GetEnumerator();
            
            // Current is set to "default(T)" if MoveNext() hasn't been called yet or if MoveNext() is called beyond
            // the last list index.
            Assert(enumerator.Current == 0);
            
            // default(int) is 0, which is the same as default(T) for my generic integer list
            Assert(default(int) == 0);
            
            int count = 0;
            while (enumerator.MoveNext())
            {
                count += enumerator.Current;
            }

            Assert(count == 12);
            Assert(enumerator.Current == 0);
            
            // The IEnumerator implementation ran out of indexes in the list to enumerate over.
            Assert(enumerator.MoveNext() == false);
            
            // The only way to iterate over the list again is to reset the IEnumerator.
            enumerator.Reset();
            bool enumeratorHasNext = enumerator.MoveNext();
            Assert(enumeratorHasNext);
            
            // To dispose of an IEnumerator after use, the "using" code block can be used.
            count = 0;
            using (enumerator)
                while (enumerator.MoveNext())
                {
                    count += enumerator.Current;
                }
            
            Assert(count == 10);
            
            // For type List, the implicit call to Dispose() at the end of the "using" block does nothing.
            Assert(enumerator.MoveNext() == false);
            
            // foreach is a syntactic shortcut for IEnumerator
            count = 0;
            foreach (var item in list)
            {
                count += item;
            }
            Assert(count == 12);
            
            // Since Queue<T> implements IEnumerable<T>, it can be passed to AndyList() as an argument.
            var enumerableQueue = new Queue<int>();
            enumerableQueue.Enqueue(8);
            enumerableQueue.Enqueue(10);
            enumerableQueue.Enqueue(12);
            
            AndyList<int> andyListFromQueue = new AndyList<int>(enumerableQueue);
            IEnumerator<int> andyListEnumerator = andyListFromQueue.GetEnumerator();

            count = 0;
            using (andyListEnumerator)
            {
                while (andyListEnumerator.MoveNext())
                {
                    count += andyListEnumerator.Current;
                }
            }
            Assert(count == 30);
            
            // Countable collections of objects implement the ICollection<T> interface along with
            // its non-generic counterpart ICollection.
            var countableList = new List<int> {2, 4};
            ICollection<int> iCollection = countableList;
            Assert(iCollection.Count == 2);

            ICollection iCollectionNonGeneric = countableList;
            Assert(iCollectionNonGeneric.Count == 2);
            
            // Collections indexed by position implement IList<T> and IList.  IList<T> extends ICollection<T>,
            // IEnumerable<T>, and IEnumerable.  IList extends ICollection and IEnumerable.
            IList<int> iList = countableList;
            IList iListNonGeneric = countableList;
            Assert(iList.IndexOf(2) == 0 && iList[1] == 4);
            Assert(iListNonGeneric.IndexOf(2) == 0 && Equals(iListNonGeneric[1], 4));

            // Setting the compile time type to IReadOnlyList<T> means I can't use methods that modify the list.
            IReadOnlyList<int> _ = countableList;
            
            IDictionary<string, int> linesWritten2019 = new Dictionary<string, int>();
            
            // Top 3 languages of the year so far.  Demonstrate the methods implemented from IDictionary.
            linesWritten2019.Add("Python", 7921);
            linesWritten2019.Add("JavaScript", 3974);
            linesWritten2019.Add("HCL", 3450);
            linesWritten2019.Add("Markdown", 2751);

            linesWritten2019.Remove("Markdown");
            
            var keys = linesWritten2019.Keys;
            var values = linesWritten2019.Values;
            Assert(values.Count == 3 && keys.Count == 3);
            Assert(linesWritten2019["Python"] == 7921);
        }
    }
}

File: Java/base/collections/CollectionInterfaces.java
package collections;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Spliterator;

/**
 * Revisiting the interfaces which make up the collections framework.
 * @author Andrew Jarombek
 * @since 7/31/2019
 */

class CollectionInterfaces {
    static void execute() {
        // Collections implement the Iterable<T> interface and specify iterator(), forEach(),
        // and spliterator() methods.  iterator() returns an instance that implements Iterator<T> which is used
        // for iterating over a collection.
        // C# - [https://bit.ly/2T4k2KN]
        List<Integer> list = List.of(2, 4, 6);
        Iterator<Integer> it = list.iterator();

        var count = 0;
        while (it.hasNext()) {
            count += it.next();
        }
        assert count == 12;

        List<Float> floatList = List.of(2.0f, 2.1f, 2.2f, 2.3f, 2.4f);
        Spliterator<Float> spliterator = floatList.spliterator();

        // Get the characteristics of a List<T> spliterator.
        int characteristics = spliterator.characteristics();
        assert characteristics == (Spliterator.ORDERED | Spliterator.SIZED | Spliterator.SUBSIZED);

        long estimatedSize = spliterator.estimateSize();
        assert estimatedSize == 5;

        int iterations = 0;
        while (spliterator.tryAdvance((item) -> System.out.println("Advancing past " + item))) {
            iterations++;
        }
        assert iterations == 5;

        // Splitting up the workload.
        var split1 = floatList.spliterator();
        var split2 = split1.trySplit();

        iterations = 0;
        while (split1.tryAdvance((item) -> System.out.println("Split #1 Advancing past " + item))) {
            iterations++;
        }
        assert iterations == 3;

        iterations = 0;
        while (split2.tryAdvance((item) -> System.out.println("Split #2 Advancing past " + item))) {
            iterations++;
        }
        assert iterations == 2;

        // Working with the custom AndyList and its inner Iterator and Spliterator classes.
        AndyList<Integer> andyList = new AndyList<>(List.of(1,2,3,4));
        Iterator<Integer> andyListIterator = andyList.iterator();

        // AndyList iterator works just like a regular List iterator.
        count = 0;
        while (andyListIterator.hasNext()) {
            count += andyListIterator.next();
        }
        assert count == 10;

        try {
            // This always throws an error
            andyListIterator.remove();
            assert false;
        } catch (UnsupportedOperationException e) {
            assert true;
        }

        // Construct a new list using the forEach() method in AndyList
        var andyListDuplicate = new ArrayList<>();
        andyList.forEach(andyListDuplicate::add);
        assert andyListDuplicate.size() == 4;

        // Append to the new list using the forEachRemaining() method in the AndyListIterator
        andyListIterator = andyList.iterator();
        andyListIterator.forEachRemaining(andyListDuplicate::add);
        assert andyListDuplicate.size() == 8;

        // Finally Java provides a syntactical shortcut for iterators with the for-each loop
        for (var item : andyList) {
            andyListDuplicate.add(item);
        }
        assert andyListDuplicate.size() == 12;

        // Test out the custom AndyListSpliterator designed for the AndyList class
        Spliterator<Integer> andyListSpliterator = andyList.spliterator();

        assert andyListSpliterator.estimateSize() == 4;
        assert andyListSpliterator.characteristics() == (Spliterator.IMMUTABLE | Spliterator.SIZED);

        iterations = 0;
        while (andyListSpliterator.tryAdvance((item) -> System.out.println("AndyList Advancing past " + item))) {
            iterations++;
        }
        assert iterations == 4;

        var andyListSplit1 = andyList.spliterator(); // mid-high (indexes 2-3)
        var andyListSplit2 = andyListSplit1.trySplit(); // low-mid (indexes 0-1)

        iterations = 0;
        while (andyListSplit1.tryAdvance((item) -> System.out.println("AndyList Split #1 Advancing past " + item))) {
            iterations++;
        }
        assert iterations == 2;

        iterations = 0;
        while (andyListSplit2.tryAdvance((item) -> System.out.println("AndyList Split #2 Advancing past " + item))) {
            iterations++;
        }
        assert iterations == 2;
    }
}
